### ISSUE TITLE
ptr_arg suggests not using `&mut Cow` in ways that are invalid or would add clones

### ISSUE THREAD

**Issue by carols10cents** [rust-lang/rust-clippy#8495]
### Summary

I have a function that takes `&mut Cow<'_, T>` and might change a `Cow::Borrowed` to a `Cow::Owned` (thus doing a clone), or it might not. Clippy suggests changing the argument to `&mut T`, which would mean always needing to call `Cow::to_mut` and unconditionally cloning.

### Lint Name

ptr_arg

### Reproducer

I tried this code:

```rust
use std::borrow::Cow;

fn main() {
    let mut c = Cow::Borrowed("hi");
    do_something_maybe(&mut c);
    assert!(matches!(c, Cow::Borrowed(..)));

    let mut b = Cow::Borrowed("bye");
    do_something_maybe(&mut b);
    assert!(matches!(b, Cow::Owned(..)));
}

fn do_something_maybe(moo: &mut Cow<'_, str>) {
    if let Some('b') = moo.chars().next() {
        moo.to_mut().push_str(", see you later");
    }
}
```

clippy suggested:

```
warning: using a reference to `Cow` is not recommended
  [--> src/main.rs:12:28
](https://play.rust-lang.org/#)   |
12 | fn do_something_maybe(moo: &mut Cow<'_, str>) {
   |                            ^^^^^^^^^^^^^^^^^ help: change this to: `&mut str`
   |
   = note: `#[warn(clippy::ptr_arg)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
```

This suggestion is invalid; if I change the code to:

```
use std::borrow::Cow;
fn main() {
    let mut c = Cow::Borrowed("hi");
    do_something_maybe(&mut c);
    assert!(matches!(c, Cow::Borrowed(..)));

    let mut b = Cow::Borrowed("bye");
    do_something_maybe(&mut b);
    assert!(matches!(b, Cow::Owned(..)));
}

fn do_something_maybe(moo: &mut str) {
    if let Some('b') = moo.chars().next() {
        moo.push_str(", see you later");
    }
}
```

then it doesn't compile:

```
error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): no method named `push_str` found for mutable reference `&mut str` in the current scope
  [--> src/main.rs:14:13
](https://play.rust-lang.org/#)   |
14 |         moo.push_str(", see you later");
   |             ^^^^^^^^ method not found in `&mut str`
```

I could change the code to this:

```
use std::borrow::Cow;
fn main() {
    let mut c = Cow::Borrowed("hi");
    do_something_maybe(c.to_mut());
    assert!(matches!(c, Cow::Borrowed(..)));

    let mut b = Cow::Borrowed("bye");
    do_something_maybe(b.to_mut());
    assert!(matches!(b, Cow::Owned(..)));
}

fn do_something_maybe(moo: &mut String) {
    if let Some('b') = moo.chars().next() {
        moo.push_str(", see you later");
    }
}
```

but that causes an extra `clone` because I have to call `to_mut` in both cases: the assertion on line 5 fails with this code.

I expected clippy to not suggest code that won't compile or that would increase the number of clones.

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: aarch64-apple-darwin
release: 1.59.0
LLVM version: 13.0.0

clippy 0.1.61 (2022-03-01 f0c4da4)
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
